hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/JoinNestedLoopPhysicalRel.java

```diff
@@ -109,6 +114,36 @@ public <V> V accept(CreateDagVisitor<V> visitor) {
         return visitor.onNestedLoopJoin(this);
     }
 
+    /**
+     * Cost calculation of Nested Loop Join relation.
+     * NLJ algorithm is simple algorithm, where for each left row we are traversing right row set.
+     * Speaking of cost estimation, we are accounting the next properties:
+     * - row count is estimating ans L * R, because for each left row we're probing full right row set;
+     * - same for CPU cost estimation multiplied by cost of row comparison.
+     * The perfect assumption also must include memory and IO cost estimation, and also
+     * a selectivity for a right row set.
+     */
+    @Override
+    @Nullable
+    public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        final double leftRowCount = mq.getRowCount(left);
+        final double rightRowCount = mq.getRowCount(right);
+        if (Double.isInfinite(leftRowCount) || Double.isInfinite(rightRowCount)) {
+            return planner.getCostFactory().makeInfiniteCost();
+        }
+
+        RelOptCost rightCost = planner.getCost(getRight(), mq);
+        if (rightCost == null) {
+            return planner.getCostFactory().makeInfiniteCost();
+        }
+
+        // TODO: introduce selectivity estimator, but ATM we taking the worst case scenario : selectivity = 1.0.
+        double rowsEstimate = leftRowCount * /* TODO: selectivity */ rightRowCount;
+        double cpuEstimate = Math.max(1.0, rowsEstimate - 1) * Cost.JOIN_ROW_CMP_MULTIPLIER;
+
+        return planner.getCostFactory().makeCost(rowsEstimate, cpuEstimate, 0);
```

Comment:
Calcite has a way of calculating the join selectivity, see `RelMdUtil.getJoinRowCount`. This number of rows is returned by `super.computeSelfCost`

Reply:
also this row count is important for calculations higher in the tree

Reply:
> Calcite has a way of calculating the join selectivity,

I'd correct you: Calcite has an **interface** to calculate the join selectivity. Standard implementation is very simple and does not cover my requirements for a selectivity estimator.
Actually, I'd like to discuss it in details.

Reply:
your current implementation of selectivity estimation is even worse than the one Calcite. If it is ready in Calcite why not use it even though it might be not perfect?